crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,295 @@
+
+        for (pk, msg) in pks.borrow().iter().zip(msgs.iter()) {
+            let mut hasher = Sha256::new();
+            hasher.update(pk);
+            hasher.update(msg); // pk + msg
+            let h: [u8; 32] = hasher.finalize().into();
+            let pairing: Option<&GTElement> = self.cache.get(&h);
+            match pairing {
+                Some(pairing) => {
+                    if !force_cache {
+                        // Heuristic to avoid more expensive sig validation with pairing
+                        // cache when it's empty and cached pairings won't be useful later
+                        // (e.g. while syncing)
+                        missing_count += 1;
+                        if missing_count > pks.borrow().len() / 2 {
+                            return vec![];
+                        }
+                    }
+                    pairings.push(Some(pairing.clone()));
+                }
+                _ => {
+                    pairings.push(None);
+                }
+            }
+        }
+
+        // G1Element.from_bytes can be expensive due to subgroup check, so we avoid recomputing it with this cache
+        let mut pk_bytes_to_g1: HashMap<[u8; 48], PublicKey> = HashMap::new();
+        let mut ret: Vec<GTElement> = vec![];
+
+        for (i, pairing) in pairings.iter_mut().enumerate() {
+            if let Some(pairing) = pairing {
+                // equivalent to `if pairing is not None`
+                ret.push(pairing.clone());
+            } else {
+                let mut aug_msg = pks.borrow()[i].to_vec();
+                aug_msg.extend_from_slice(msgs[i].as_ref()); // pk + msg
+                let aug_hash: Signature = hash_to_g2(&aug_msg);
+
+                let pk_parsed: &mut PublicKey = pk_bytes_to_g1
+                    .entry(pks.borrow()[i])
+                    .or_insert_with(|| PublicKey::from_bytes(&pks.borrow()[i]).unwrap());
+
+                let pairing: GTElement = aug_hash.pair(pk_parsed);
+                let mut hasher = Sha256::new();
+                hasher.update(&aug_msg);
+                let h: [u8; 32] = hasher.finalize().into();
+                self.cache.put(h, pairing.clone());
+                ret.push(pairing);
+            }
+        }
+
+        ret
+    }
+
+    pub fn aggregate_verify(
+        &mut self,
+        pks: &Vec<[u8; 48]>,
+        msgs: &[&[u8]],
+        sig: &Signature,
+        force_cache: bool,
+    ) -> bool {
+        let mut pairings: Vec<GTElement> = self.get_pairings(pks, msgs, force_cache);
+        if pairings.is_empty() {
+            let mut data = Vec::<(PublicKey, &[u8])>::new();
+            for (pk, msg) in pks.iter().zip(msgs.iter()) {
+                let pk = PublicKey::from_bytes_unchecked(pk).unwrap();
+                data.push((pk.clone(), msg));
+            }
+            return agg_ver(sig, data);
+        }
+        let pairings_prod = pairings.pop(); // start with the first pairing
```

Comment:
at this point you're guaranteed that `pairings` is not empty, right? So it would seem you could just do:
```
let pairing_prod = pairing.pop().expect("internal error");
```

But I also think that this loop will be quite slow compared to the built-in BLST functions to verify a signature.

See this function: https://github.com/Chia-Network/chia_rs/blob/main/crates/chia-bls/src/signature.rs#L403